ddmark/validation.go

```diff
@@ -151,33 +199,102 @@ func (l LinkedFields) ApplyRule(fieldvalue reflect.Value) error {
 	}
 
 	for _, item := range l {
-		if structMap[item] != nil {
+		res, err := checkValueExistsOrIsValid(item, structMap, ruleName(l))
+		if err != nil {
+			return err
+		}
+
+		if res {
 			matchCount++
 		}
 	}
 
 	if matchCount != 0 && matchCount != len(l) {
-		return fmt.Errorf("%v: all of the following fields need to be either nil of non-nil (currently unmatched): %v", ruleName(l), l)
+		return l.GenValueCheckError()
 	}
 
 	return nil
 }
 
-func (r AtLeastOneOf) ApplyRule(fieldvalue reflect.Value) error {
+func (l LinkedFieldsValue) GenValueCheckError() error {
+	template := "%v: all of the following fields need to be either nil/at the indicated value or non-nil/not at the indicated value; currently unmatched: %v"
+	return fmt.Errorf(template, ruleName(l), l)
+}
+
+func (l LinkedFieldsValueWithTrigger) ApplyRule(fieldvalue reflect.Value) error {
 	fieldvalue = reflect.Indirect(fieldvalue)
 
+	var matchCount = 0
+	// room for logic to possibly expand the marker to accept multiple/combined trigger values (instead of 1)
+	var c = 1
+
+	if len(l) < 2 {
+		return fmt.Errorf("%v: marker was wrongly defined in struct: less than 2 fields", ruleName(l))
+	}
+
 	structMap, ok := structValueToMap(fieldvalue)
 	if !ok {
-		return fmt.Errorf("%v: marker applied to wrong type: currently %v, can only be %v", ruleName(r), fieldvalue.Type(), "struct")
+		return fmt.Errorf("%v: marker applied to wrong type: currently %v, can only be %v", ruleName(l), fieldvalue.Type(), "struct")
 	}
 
-	for _, item := range r {
+	for _, markerString := range l[:c] {
+		res, err := checkValueExistsOrIsValid(markerString, structMap, ruleName(l))
+		if err != nil {
+			return err
+		}
+
+		if res {
+			matchCount++
+		}
+	}
+
+	if matchCount != len(l[:c]) {
+		return nil
+	}
+
+	for _, markerString := range l[c:] {
+		res, err := checkValueExistsOrIsValid(markerString, structMap, ruleName(l))
+		if err != nil {
+			return err
+		}
+
+		if res {
+			matchCount++
+		}
+	}
+
+	if matchCount != 0 && matchCount != len(l) {
+		return l.GenValueCheckError()
+	}
+
+	return nil
+}
+
+func (l LinkedFieldsValueWithTrigger) GenValueCheckError() error {
+	template := "%v: all of the following fields need to be aligned; if the first value is valid / exists, all the following need to either exist or have the indicated value: %v"
```

Comment:
Can we possibly extract "the first value" out from `l` here, so it looks like:
```
all of the following fields need to be aligned; if nodeFailure is set, all the following need to either exist or have the indicated value: [level]
```

I think "valid / exists" might be too vague, we can probably assume that if we got here, we're dealing with something valid (or at least something parseable). These messages are primarily for inexperienced users, after all

Reply:
good point; I'll adapt that now

Reply:
done !